runtime-sdk/src/modules/core/mod.rs

```diff
@@ -453,22 +468,89 @@ impl<Cfg: Config> Module<Cfg> {
             extra_gas += params.gas_costs.auth_signature;
         }
 
-        let propagate_failures = args.propagate_failures;
-        ctx.with_simulation(|mut sim_ctx| {
-            sim_ctx.with_tx(0 /* index */, tx_size, args.tx, |mut tx_ctx, call| {
-                let (result, _) =
-                    dispatcher::Dispatcher::<C::Runtime>::dispatch_tx_call(&mut tx_ctx, call);
-                if propagate_failures && !result.is_success() {
-                    // Report failure.
-                    let err: TxSimulationFailure = result.try_into().unwrap(); // Guaranteed to be a Failed CallResult.
-                    return Err(Error::TxSimulationFailed(err));
-                }
-                // Don't report success or failure. If the call fails, we still report
-                // how much gas it uses while it fails.
-                let gas_used = *tx_ctx.value::<u64>(CONTEXT_KEY_GAS_USED).or_default();
-                Ok(gas_used + extra_gas)
+        // Simulates a transaction.
+        let mut simulate = |tx: transaction::Transaction, report_failure: bool| {
+            ctx.with_simulation(|mut sim_ctx| {
+                sim_ctx.with_tx(0 /* index */, tx_size, tx, |mut tx_ctx, call| {
+                    let (result, _) =
+                        dispatcher::Dispatcher::<C::Runtime>::dispatch_tx_call(&mut tx_ctx, call);
+                    if !result.is_success() && report_failure {
+                        // Report failure.
+                        let err: TxSimulationFailure = result.try_into().unwrap(); // Guaranteed to be a Failed CallResult.
+                        return Err(Error::TxSimulationFailed(err));
+                    }
+                    // Don't report success or failure. If the call fails, we still report
+                    // how much gas it uses while it fails.
+                    let gas_used = *tx_ctx.value::<u64>(CONTEXT_KEY_GAS_USED).or_default();
+                    Ok(gas_used)
+                })
             })
-        })
+        };
+
+        // Do a binary search for exact gas limit.
```

Comment:
It would be great if we could increase unit test coverage of the below if not too hard.

Reply:
Yeah I think it shouldn't be too hard.